src/lib/mina_lib/mina_lib.ml

```diff
@@ -1495,6 +1495,126 @@ let start_filtered_log ~commit_id
       () ;
     Ok () )
 
+let fetch_completed_snarks (module Context : CONTEXT) snark_pool network
+    top_block get_most_recent_valid_block =
+  let open Context in
+  let open Network_peer in
+  let%bind all_peers = Mina_networking.peers network in
+  Deferred.List.iter
+    ~f:(fun peer ->
+      [%log debug] "PEER IS: Fetching completed snarks from peer: $peer"
+        ~metadata:[ ("peer", Network_peer.Peer.to_yojson peer) ] ;
+      let completed_works =
+        Mina_networking.get_completed_checked_snarks network peer
+      in
+      let%bind completed_works = completed_works in
+      let completed_works =
+        match completed_works with
+        | Error e ->
+            [%log debug]
+              ~metadata:
+                [ ("peer", Network_peer.Peer.to_yojson peer)
+                ; ("error", Error_json.error_to_yojson e)
+                ]
+              "Failed to fetch completed snarks from peer: $error" ;
+            []
+        | Ok completed_works ->
+            completed_works
+      in
+
+      [%log debug]
+        ~metadata:
+          [ ("peer", Network_peer.Peer.to_yojson peer)
+          ; ("completed_works", `Int (List.length completed_works))
+          ]
+        "Fetched $completed_works completed snarks from peer: $peer" ;
+
+      (* keep reading from the frontier pipe until the new top block is greater than 2 more than the old top block.public_key
+         we do this to account for the block lag placed after the bootstrap process. The number 2 was chosen after manual testing.
+         1 did not yield successful verifcation of the polled snarks work.
+      *)
+      let rec wait_for_new_top_block old_top_block =
```

Comment:
Any reason for this to be inside the loop? if not, slightly prefer this outside to make the logic a bit more readable i.e, the node gets the almost latest block state and then queries peers for snark work.

Reply:
The alternative is storing the rpc response in memory for completed work and processing once the best tip is up to date. You cannot query the best tip and verify at the same time due to the sync delay. This was imo the most memory efficient solution. I'm made a decision to trade the time it takes in the verification loop for minimal memory usage.

Reply:
I meant to move just the wait for best tip and not have to do it for per peer. Verification should be done per query response.

Reply:
another reason it is in the loop is so we can query peers as soon as we get the most recent valid tip. it is an attempt to grab data as soon as possible so the network doesn't advance while we are attempting to sync the transition frontier.